sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -3286,101 +3283,91 @@ pi_result piQueueGetInfo(pi_queue Queue, pi_queue_info ParamName,
 }
 
 pi_result piQueueRetain(pi_queue Queue) {
-  // Lock automatically releases when this goes out of scope.
-  std::scoped_lock lock(Queue->Mutex);
   Queue->RefCountExternal++;
-  piQueueRetainNoLock(Queue);
+  piQueueRetainInternal(Queue);
   return PI_SUCCESS;
 }
 
 pi_result piQueueRelease(pi_queue Queue) {
   PI_ASSERT(Queue, PI_INVALID_QUEUE);
-  {
-    // Have this scope such that the lock is released before the
-    // queue is potentially deleted in QueueRelease.
-    // Lock automatically releases when this goes out of scope.
-    std::scoped_lock lock(Queue->Mutex);
 
-    Queue->RefCountExternal--;
-    if (Queue->RefCountExternal == 0) {
-      // When external reference count goes to zero it is still possible
-      // that internal references still exists, e.g. command-lists that
-      // are not yet completed. So do full queue synchronization here
-      // and perform proper cleanup.
-      //
-      // It is possible to get to here and still have an open command list
-      // if no wait or finish ever occurred for this queue.
-      if (auto Res = Queue->executeAllOpenCommandLists())
-        return Res;
+  // Only single thread can reach external ref count equal to 0. So, code under
+  // if condition doesn't need to be guarded with a mutex. If external ref count
+  // reaches 0 then it means that queue is not accessible anymore from outside
+  // of the Level Zero plugin. Behavior is undefined if DPCPP runtime calls
+  // piQueueRelease on a queue with external ref count 0 or provides such queue
+  // as an argument to any of the pi functions.
+  if (--(Queue->RefCountExternal) == 0) {
```

Comment:
nit: I'd personally use `fetch_sub` interface as not every reader is aware (I certainly wasn't one of them) that `operator--` provides the guarantees and it's more likely they'll "cppreference" the member function than an overloaded operator.